azkaban-common/src/main/java/azkaban/version/JdbcVersionSetLoaderImpl.java

```diff
@@ -0,0 +1,217 @@
+    private static final Logger logger = Logger.getLogger(JdbcVersionSetLoaderImpl.class);
+
+    private static final String SELECT_VSET_FROM_NUM = "SELECT vset.num, vset.md5, vset.json FROM version_set vset WHERE vset.num=?";
+    private static final String SELECT_VSET_FROM_MD5 = "SELECT vset.num, vset.md5, vset.json FROM version_set vset WHERE vset.md5=?";
+    private static final String SELECT_ALL_VSET = "SELECT vset.num, vset.md5, vset.json FROM version_set vset";
+    private static final String INSERT_VSET = "INSERT INTO version_set (md5, json) values (?,?)";
+    private static final String DELETE_VSET = "DELETE FROM version_set vset WHERE vset.md5=?";
+
+    private final DatabaseOperator dbOperator;
+    private final Map<String, VersionSet> md5ToVersionSet = new HashMap<>();
+    private final Map<Integer, VersionSet> numToVersionSet = new HashMap<>();
+
+    /**
+     * This constructor should be instantiated using dependency injection only
+     * to make sure that that class remains Singleton.
+     *
+     * @param dbOperator Interface to perform db related operations.
+     */
+    @Inject
+    public JdbcVersionSetLoaderImpl(DatabaseOperator dbOperator) throws IOException {
+        this.dbOperator = dbOperator;
+        for (VersionSet versionSet : fetchAllVersionSets()) {
+            this.md5ToVersionSet.put(versionSet.getVersionSetMd5Hex(), versionSet);
+            this.numToVersionSet.put(versionSet.getVersionSetNum(), versionSet);
+        }
+    }
+
+    /**
+     * This method inserts the versionSetJsonString and versionSetMd5Hex to the Table version_set
+     * and get the versionSetNum autogenerated post insertion.
+     */
+    @Override
+    public synchronized int insertAndGetVersionSetNum(String versionSetJsonString, String versionSetMd5Hex) throws IOException {
+        final SQLTransaction<Integer> insertVersionSet = transOperator -> {
+            final VersionSetHandler versionSetHandler = new VersionSetHandler();
+            // Check if the versionSet already exists for versionSetMd5Hex
+            List<VersionSet> versionSets = transOperator.query(SELECT_VSET_FROM_MD5, versionSetHandler, versionSetMd5Hex);
+            if (versionSets.size() > 1) {
+                throw new SQLException("Expected only one VersionSet for newly inserted versionSetMd5Hex " + versionSetMd5Hex);
+            }
+            // Insert if it doesn't
+            if (versionSets.isEmpty()) {
+                transOperator.update(INSERT_VSET, versionSetMd5Hex, versionSetJsonString);
+            }
+
+            // Query for versionSet again after insert
+            versionSets = transOperator.query(SELECT_VSET_FROM_MD5, versionSetHandler, versionSetMd5Hex);
+            if (versionSets.size() > 1) {
+                throw new SQLException("Expected only one VersionSet for newly inserted versionSetMd5Hex " + versionSetMd5Hex);
+            }
+            if (versionSets.isEmpty()) {
```

Comment:
Empty should not be a valid case. exception should be thrown like I mentioned in above comment. Can you please follow https://github.com/azkaban/azkaban/blob/1e8c84c13d20af27bcd47da5d42040d1ec996a00/azkaban-common/src/main/java/azkaban/executor/ExecutionFlowDao.java#L82 and write insert related code using transOperator.update method? You will get id in return and won't need to get version set again. It will save a query.

Reply:
Done